cube/generic_networks/att_model.py

```diff
@@ -0,0 +1,215 @@
+		self.vocab_size = vocab_size + 1
+
+	def fit(self, X_train=None, y_train=None, X_test=None, y_test=None, epochs=None):
+		X_train, y_train, X_test, y_test = self._prepare_data(X_train, y_train, X_test, y_test)
+		self._initialize_model()
+
+		print('Starting training...')
+
+		self.trainer = dy.AdamTrainer(self.model)
+		losses = []
+
+		for _ in tqdm(range(epochs)):
+			curr_loss = 0.0
+
+			for X, y in zip(X_train, y_train):
+				y_prob = self._predict_proba(X)
+
+				loss = dy.sum_batches(dy.pickneglogsoftmax_batch(y_prob, y))
+				curr_loss += loss.value()
+
+				loss.backward()
+				self.trainer.update()
+
+			losses.append(curr_loss / len(X_train))
+
+			print('Train Loss:', losses[-1])
+			self.evaluate(X_test, y_test)
+			print()
+
+		print('Done training')
+
+	def evaluate(self, X_test=None, y_test=None):
+		acc = []
+		curr_loss = 0.0
+
+		for X, y in zip(X_test, y_test):
+			proba = self._predict_proba(X, train=False)
+			curr_loss += dy.sum_batches(dy.pickneglogsoftmax_batch(proba, y)).value()
+
+			proba = proba.npvalue()
+
+			for i in range(len(proba[0])):
+				y_pred = argmax(proba[:, i])
+				y_true = y[i]
+
+				acc.append(1 if y_pred == y_true else 0)
+
+		curr_loss /= len(X_test)
+
+		print('Validation Loss:', curr_loss)
+		print('Test Accuracy:', mean(acc))
+
+
+	"""
+		Defining the model's trainable parameters. We use one layer of LSTM/GRUs to process
+		the sequences, on top of which we place a predictive attention mechanism to emphasize
+		the words that are most relevant. The resulting context vector, which acts like a summary
+		is projected, via multiplication with a learnable matrix, to the label space.
+	"""
+	def _initialize_model(self):
+		self.model = dy.Model()
+		self.input_lookup = self.model.add_lookup_parameters((self.vocab_size, self.embedding_size))
+
+		# Attention params
+		self.attention_w1 = self.model.add_parameters((self.state_size, self.state_size))
+		self.attention_w2 = self.model.add_parameters((self.state_size, self.state_size))
+		self.attention_v = self.model.add_parameters((1, self.state_size))
+
+		# Predictive allignment params
+		self.w_p = self.model.add_parameters((self.state_size, self.state_size))
+		self.v_p = self.model.add_parameters((1, self.state_size))
+
+		# LSTM/GRU and last layer projection matrix
+		self.lstm = dy.GRUBuilder(self.lstm_num_of_layers, 
+										self.embedding_size, self.state_size, self.model)
+
+		self.output_w = self.model.add_parameters((self.num_of_classes, self.state_size))
+		self.output_b = self.model.add_parameters((self.num_of_classes))
+
+	"""
+		Predictive allignments based on importance of words to each class. Here W_p and v_p are
+		learnable parameters that get adjusted during training.
+	"""
+	def _predictive_allignment(self, sent_len, input_vector):
+		w_p = dy.parameter(self.w_p)
+		v_p = dy.parameter(self.v_p)
+
+		return sent_len * dy.logistic(v_p * dy.tanh(w_p * input_vector))
+
+	def _get_allignments(self, input_vectors):
```

Comment:
Do you need to back propagate through this computation? If so, you should use DyNET primitives instead of numpy.